### ISSUE TITLE
Upload to conda broken?

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#438]
Looking at https://conda.anaconda.org/omnia, it appears that the OPS-dev hasn't been updated since Jan 24 (merge of #406). I don't see any reason for that (or the next merge, #431, or the previous merge, #428) to have affected anything about the conda upload process.

However, in the most recent build, merging #429, I see `[Unauthorized] (u'Invalid Token', 401)` (line 2739: https://travis-ci.org/choderalab/openpathsampling/builds/113427068#L2739). Has our token somehow changed?

**Comment by jhprinz** [comment]
Maybe @jchodera knows about this?

**Comment by jchodera** [comment]
It may simply have expired. I can fix.

**Comment by jchodera** [comment]
Actually, would you mind if I switched to nightly dev builds for both osx and linux? Or do you need a push to conda every commit?

**Comment by dwhswenson** [comment]
We upload to conda on every commit to master, which is (usually) less frequently than nightly. I wouldn't mind doing nightly builds (in fact, I'm currently trying out nightli.es for nightly test builds on other projects, and was going to suggest it in a week or two if it kept being as useful as it has been -- the *lack* of errors from my other projects is what alerted me to this problem).

So as far as I'm concerned, either approach is good!

**Comment by jhprinz** [comment]
I guess for me. How would nightly builds be different from what we have now? Every change to master is build on travis and then pushed to conda. 

Does nightly also include builds for other branches than master? Or do we have a separate master-dev branch?

**Comment by jchodera** [comment]
The simplest fix is to add the recipe to https://github.com/omnia-md/conda-dev-recipes/ via a PR.

Those are built nightly for both `osx` and `linux` (using the same Docker build framework we use for https://github.com/omnia-md/conda-recipes).

**Comment by dwhswenson** [comment]
@jhprinz : If I understand, the main advantage here is to make it so that we don't have to directly manage the conda upload or related tokens -- @jchodera takes on responsibility for that (and only has to do any updates once to fix all the projects in the dev-recipes repo).

There is a difference in timing, so there's up to a 12-hour window during which the conda version is behind the master branch, instead of it immediately updating after merging a PR. This means that I can't change something in OPS, merge the PR, use those changes in `dynamiq_engine`, and then expect `dynamiq_engine` to pass its Travis tests (until the next day, when conda catches up.)

That's a disadvantage, but not a significant one.

If the `omnia-md/conda-dev-recipes` approach sounds like a good idea to you, JHP, then I'll fork that repo and put together a PR to add OPS to it (probably sometime this weekend).

**Comment by jhprinz** [comment]
Thanks, that explains it and totally makes sense to do. The timelag between pushing and build does not matter. A developer can always checkout any branch and install manually if he needs it quickly.

Please go ahead.

**Comment by jchodera** [comment]
I'm on it!

https://github.com/omnia-md/conda-dev-recipes/pull/29

**Comment by dwhswenson** [comment]
Well, since you're already half (3/4? 90%?) done, I'll switch the assignment to you! :-)

**Comment by jchodera** [comment]
Is `openpathsampling` not python 3.x compatible?
```

Package: openpathsampling-dev-0.0.0-py34_0

source tree in: /anaconda/conda-bld/work

+ /anaconda/envs/_build/bin/python setup.py install

Traceback (most recent call last):

  File "setup.py", line 148, in <module>

    main()

  File "setup.py", line 144, in main

    setupKeywords = buildKeywordDictionary()

  File "setup.py", line 133, in buildKeywordDictionary

    for key in sorted( setupKeywords.iterkeys() ):

AttributeError: 'dict' object has no attribute 'iterkeys'

Command failed: /bin/bash -x -e /io/openpathsampling-dev/build.sh
```

**Comment by dwhswenson** [comment]
> Is openpathsampling not python 3.x compatible?

Nope. Our plan has been to get a feature-complete 1.0, then add python 3.x support. #70.

**Comment by jchodera** [comment]
Done: https://github.com/omnia-md/conda-dev-recipes/pull/29

Can you remove the conda upload from `.travis.yml`?

**Comment by jhprinz** [comment]
I will quickly do that.

**Comment by jhprinz** [comment]
I will leave updating the docs? Or are these also taken care of by the nightly build?

**Comment by jchodera** [comment]
You'll still need to have the docs pushed by `.travis.yml`.

**Comment by dwhswenson** [comment]
According to https://conda.anaconda.org/omnia, we've updated (nearly 12 hours ago), but I'm still seeing old versions in practice.

This is easy to see in https://github.com/dynamiq-md/dynamiq_engine/pull/19, where, right before the nosetests, I report the full OPS version (https://travis-ci.org/dynamiq-md/dynamiq_engine, line 633). It gives commit 056e44a (Jan 24).

Should I just wait a day and see if there's some caching somewhere that needs to be cleared, or is this a genuine problem?

**Comment by jchodera** [comment]
> According to https://conda.anaconda.org/omnia, we've updated (nearly 12 hours ago), but I'm still seeing old versions in practice.

Clean your local package cache first:
```
conda clean -pltis --yes
```

**Comment by dwhswenson** [comment]
> Clean your local package cache first:

Where should that be? And why would that have any effect? I put it after conda gets installed though the `conda_install.sh` script. No difference. https://travis-ci.org/dynamiq-md/dynamiq_engine/builds/114023415 (now line 690). I tried again by putting it just before my package gets built. No difference. https://travis-ci.org/dynamiq-md/dynamiq_engine/builds/114024218 (line 716). I tried again by putting it after switching to the test environment. No difference https://travis-ci.org/dynamiq-md/dynamiq_engine/builds/114024651 (line 812).

I wouldn't have expected that to have any effect anyway. After all, isn't Travis building a clean conda install every time it runs? How can there be old cached packages on Travis?

**Comment by jchodera** [comment]
Sorry, I hadn't realized you were talking about Travis. What is dynamiq-md?

**Comment by jchodera** [comment]
I think I have to remove the older dev packages from anaconda cloud manually because I changed the version number to 0.0.0

**Comment by dwhswenson** [comment]
> What is dynamiq-md?

Side project of mine. It builds off of some of the OPS trajectory stuff, but adds additional features required for semiclassical dynamics (the "q" is for "quantum effects").

Is it possible that version numbers are causing the problem? We variously have this labelled as 0.0.3, 0.0.1a0, and 0.0.0. (In particular, looking at the log at https://s3.amazonaws.com/archive.travis-ci.org/jobs/113957395/log.txt [from the conda-dev-recipes] it looks like we're building into `site-packages/openpathsampling-0.0.1a0-py2.7.egg`, although we call this 0.0.3.) I can't find any evidence of a change,  that would have involved those version numbers, but I could imagine that if conda thought the old commit was the last 0.0.3 and the new stuff is 0.0.1, it would default to serving the newer version.

... I'll try fixing the conda version in `dynamiq_engine` to 0.0.1 and see if that has any effect.

**Comment by dwhswenson** [comment]
It looks like setting the OPS version to 0.0.0 in my meta.yaml is giving the result I expected (`dynamiq_engine` fails to build, but for the right reason).

That works for now; I expect that before long we'll be giving OPS a version number large enough to imply "suitable for public consumption," at which point that will be a higher number than anything in the anaconda archives.

**Comment by jchodera** [comment]
I've deleted non-0.0.0 versions of `openpathsampling-dev` from anaconda cloud:
https://anaconda.org/omnia/openpathsampling-dev/files

You should be safe to remove the 0.0.0 constraint.

> Side project of mine. It builds off of some of the OPS trajectory stuff, but adds additional features required for semiclassical dynamics (the "q" is for "quantum effects").

Does that mean that there is lots of code duplication, or are you breaking out shared components into other repos?

**Comment by dwhswenson** [comment]
> I've deleted non-0.0.0 versions of openpathsampling-dev from anaconda cloud

Thanks -- I'll try that in an upcoming PR (as I'm fixing `dynamiq_engine` to work with the newest snapshot setup).

> Does that mean that there is lots of code duplication, or are you breaking out shared components into other repos?

*Short version*: The whole point is to have as much shared code as possible.

*Slightly longer version (OPS-related)*: @jhprinz and I have been working on making some aspects (the engine wrappers in particular, as well as JHP's netcdfplus storage tricks) easily separable from the rest of the OPS code -- these are really cool and widely reusable packages. I think the roadmap is to split these into separate repos in OPS v.2.0. I'm already reusing them in the dynamiq-md projects just by importing all of OPS.

*Much longer version (dynamiq-related)*: Semiclassical dynamics comes in a few varieties. The more accurate versions require the second derivatives of the potential (hessians), as well as the first derivatives (forces). Hence the need for a different MD engine that is better suited to those needs (`dynamiq_engine`). Less accurate versions do not need the extra derivatives. To maximize code reuse, I want to implement one engine that can handle either. @jhprinz's recent work on snapshot features should make that easier to do.

The whole dynamiq-md project will eventually include library layer repos (`dynamiq_engine`, `dynamiq_samplers`, and `spectral_densities`, if not more) and application layer repos (`pyWigner` and `dynamiq` at the least). `dynamiq` does calculations that require 2nd derivatives, and so requires `dynamiq_engine`. However, the methods in `pyWigner` only require first derivatives. Because our of `engines` API, it will be really easy for `pyWigner` to use OpenMM instead of `dynamiq_engine`.

Furthermore, there's reason to think that a shooting approach might be useful to generate trajectories in semiclassical dynamics. I explored this a bit during my PhD, but found only a negligible (~1%) speedup. However, I've since realized that I was testing in a worst-case scenario, and I should try again with a different system. By having dynamiq-md already compatible with OPS, it will be much easier to use the OPS path movers to create trial trajectories for dynamiq.

Also, I'm planning to cythonize `dynamiq_engine`. Since it will have some simple 2D models similar to the ToyEngine, the plan is that ToyEngine remains a pedagogical example of what an MD engine looks like, and how to interface one with OPS. However, real research on simple models would probably use the versions based on the optimized `dynamiq_engine`.

I've been meaning to make a picture showing the interrelatedness of all these projects (I showed JHP a sketch over Skype a few weeks ago) -- I think that picture will tell the story much more clearly. The overall point is that I see all of these as more of an ecosystem of related packages than as separate projects.
